java/ql/src/experimental/Security/CWE/CWE-094/JexlInjectionLib.qll

```diff
@@ -0,0 +1,464 @@
+      m instanceof JxltEngineCreateExpressionMethod or
+      m instanceof UnifiedJexlParseMethod or
+      m instanceof JexlEngineCreateExpressionMethod
+    ) and
+    ma.getAnArgument().getType() instanceof TypeString and
+    ma.getAnArgument() = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl expression.
+ */
+predicate creatingTaintedJxltEngineExpression(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (m instanceof JxltEngineCreateExpressionMethod or m instanceof UnifiedJexlParseMethod) and
+    ma.getAnArgument().getType() instanceof TypeString and
+    ma.getAnArgument() = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl template.
+ */
+predicate creatingTaintedJexlTemplate(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (m instanceof JxltEngineCreateTemplateMethod or m instanceof UnifiedJexlCreateTemplateMethod) and
+    (
+      isCreateTemplateSourceArg(ma, 0, node1.asExpr()) or
+      isCreateTemplateSourceArg(ma, 1, node1.asExpr())
+    ) and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if:
+ * - `expr` is an argument with the `index`
+ * - `expr` is a string or an instance of `Reader`
+ */
+predicate isCreateTemplateSourceArg(MethodAccess ma, int index, Expr expr) {
+  (
+    ma.getArgument(index).getType() instanceof TypeString or
+    ma.getArgument(index).getType() instanceof Reader
+  ) and
+  ma.getArgument(index) = expr
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that creates a Jexl script.
+ */
+predicate creatingTaintedJexlScript(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    m instanceof JexlEngineCreateScriptMethod and
+    ma.getArgument(0).getType() instanceof TypeString and
+    ma.getArgument(0) = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step
+ * that creates a callable from a Jexl expression or script.
+ */
+predicate creatingTaintedJexlCallable(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    (m instanceof JexlExpressionCallableMethod or m instanceof JexlScriptCallableMethod) and
+    ma.getQualifier() = node1.asExpr() and
+    node2.asExpr() = ma
+  )
+}
+
+/**
+ * Holds if `node1` to `node2` is a dataflow step that returns data from
+ * a tainted bean by calling one of its getters.
+ */
+predicate returningTaintedDataFromBean(DataFlow::Node node1, DataFlow::Node node2) {
+  exists(MethodAccess ma, Method m | ma.getMethod() = m |
+    m instanceof GetterMethod and
+    ma.getQualifier() = node1.asExpr() and
+    ma = node2.asExpr()
+  )
+}
+
+/**
+ * Holds if `expr` is a call to one of the methods that execute a Jexl script.
```

Comment:
```suggestion
 * Holds if `expr` calls one of the methods that execute a Jexl script against qualifier `expr`.
```
(because `expr` isn't the call expression, but the qualifier)